src/com/perl5/lang/perl/idea/refactoring/extract/PerlExtractMethodHandler.java

```diff
@@ -0,0 +1,139 @@
+
+    if (selectedText == null || selectedText.isEmpty()) {
+      return;
+    }
+
+    PerlRefactor perlRefactor = new PerlRefactor(selectedText, "myNewMethod");
+
+    Set<String> parameterVariables = new ArrayListSet<>();
+    parameterVariables.addAll(perlRefactor.getParameters());
+
+    Set<String> returnVariables = new ArrayListSet<>();
+    returnVariables.addAll(perlRefactor.getInnerVariables());
+
+    // TODO: create PerlCodeFragment that contains my(...) = foo(bar);
+    // TODO: figure how to use returnInside
+    CodeFragment fragment = new CodeFragment(parameterVariables, returnVariables,false);
+
+    // This populates the parameters and signature
+    final ExtractMethodDecorator decorator = (methodName, variableData) -> {
+
+      StringBuilder builder = new StringBuilder();
+      boolean first = true;
+
+      if (perlRefactor.getOuterVariablesModified().size() > 0) {
+        builder.append("my (");
+        for (String ov : perlRefactor.getOuterVariablesModified()) {
+          if (first) {
+            first = false;
+          }
+          else {
+            builder.append(", ");
+          }
+          builder.append(ov);
+        }
+        builder.append(") = ");
+        first = true;
+      }
+
+      builder.append(methodName).append("(");
+      for (AbstractVariableData vd : variableData) {
+        if (!vd.passAsParameter) continue;
+        if (first) {
+          first = false;
+        }
+        else {
+          builder.append(", ");
+        }
+        builder.append(vd.name);
+      }
+      builder.append(");");
+
+      return builder.toString();
+    };
+
+    //This validator should check if name will clash with existing methods
+    final ExtractMethodValidator validator = new ExtractMethodValidator() {
+      // TODO: Check for existing methods
+      @Nullable
+      @Override
+      public String check(String name) {
+        return null;
+      }
+      // TODO: Validate name
+      @Override
+      public boolean isValidName(String name) {
+        return true;
+      }
+    };
+
+    final AbstractExtractMethodDialog dialog = new AbstractExtractMethodDialog(project, "myNewMethod",
+                                                                               fragment, validator, decorator,
+                                                                               PerlFileTypeScript.INSTANCE);
+    dialog.show();
+
+    if (dialog.isOK()) {
+      String methodName = dialog.getMethodName();
+      AbstractVariableData[] variableData = dialog.getAbstractVariableData();
+      final PerlRefactorUtils refactoring = new PerlRefactorUtils(file, selectedText, methodName, selectionStartLine, selectionLength);
+
+      String methodSignature = decorator.createMethodSignature(methodName, variableData);
+
+      String method = perlRefactor.buildMethod(methodName, variableData, perlRefactor.getCodeSelection(), perlRefactor.buildMethodReturn());
```

Comment:
It's strange to call method in `perlRefactor` and pass two arguments got from `perlRefactor`

Reply:
agreed